src/problems/MetalAdvectionProblem/test_sne.cpp

```diff
@@ -0,0 +1,814 @@
+	const Real stddev = hscale / geom[0].ProbLength(2) / 2.;
+
+	const Real expectation_value = ks_sigma_sfr * domain_area * dt_coarse;
+
+	const int count = static_cast<int>(amrex::RandomPoisson(expectation_value));
+
+	// resize particle arrays
+	amrex::Array<int, 1> const lo{0};
+	amrex::Array<int, 1> const hi{count};
+	userData_.blast_x = std::make_unique<amrex::TableData<Real, 1>>(lo, hi, amrex::The_Pinned_Arena());
+	userData_.blast_y = std::make_unique<amrex::TableData<Real, 1>>(lo, hi, amrex::The_Pinned_Arena());
+	userData_.blast_z = std::make_unique<amrex::TableData<Real, 1>>(lo, hi, amrex::The_Pinned_Arena());
+	userData_.nblast = count;
+	userData_.SN_counter_cumulative += count;
+
+	// for each, sample location at random
+	auto const &px = userData_.blast_x->table();
+	auto const &py = userData_.blast_y->table();
+	auto const &pz = userData_.blast_z->table();
+	for (int i = 0; i < count; ++i) {
+		px(i) = geom[0].ProbLength(0) * amrex::Random();
+		py(i) = geom[0].ProbLength(1) * amrex::Random();
+		pz(i) = geom[0].ProbLength(2) * amrex::RandomNormal(mean, stddev);
+	}
+}
+
+/*******************************************************************/
+
+template <> void QuokkaSimulation<NewProblem>::computeAfterLevelAdvance(int lev, amrex::Real time, amrex::Real dt_lev, int ncycle)
+{
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_lo = geom[lev].ProbLoArray();
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> prob_hi = geom[lev].ProbHiArray();
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> const &dx = geom[lev].CellSizeArray();
+
+	AddSupernova(state_new_cc_[lev], prob_lo, prob_hi, dx, userData_, lev);
+}
+
+template <>
+AMREX_GPU_DEVICE AMREX_FORCE_INLINE auto
+HydroSystem<NewProblem>::GetGradFixedPotential(amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> posvec) -> amrex::GpuArray<amrex::Real, AMREX_SPACEDIM>
+{
+
+	amrex::GpuArray<amrex::Real, AMREX_SPACEDIM> grad_potential;
+
+	double x = posvec[0];
+
+	grad_potential[0] = 0.0;
+
+#if (AMREX_SPACEDIM >= 2)
+	double y = posvec[1];
+	grad_potential[1] = 0.0;
+#endif
+#if (AMREX_SPACEDIM >= 3)
+	double z = posvec[2];
+
+	// Interpolate to find the accurate g-value from array-- because linterp doesn't work on Setonix
+	size_t i = 0;
+	double x_interp = std::abs(z);
+	amrex::GpuArray<amrex::Real, 64> xx = z_data;
+	amrex::GpuArray<amrex::Real, 64> yy = logg_data;
+	while (i < xx.size() - 1 && x_interp > xx[i + 1]) {
+		i++;
+	}
+
+	// Perform linear interpolation
+	const Real x1 = xx[i];
+	const Real x2 = xx[i + 1];
+	const Real y1 = yy[i];
+	const Real y2 = yy[i + 1];
+
+	amrex::Real ginterp = (y1 + (y2 - y1) * (x_interp - x1) / (x2 - x1));
+
+	grad_potential[2] = 2. * 3.1415 * Const_G * rho_dm * std::pow(R0, 2) * (2. * z / std::pow(R0, 2)) / (1. + std::pow(z, 2) / std::pow(R0, 2));
+	grad_potential[2] += 2. * 3.1415 * Const_G * Sigma_star * (z / z_star) * (std::pow(1. + z * z / (z_star * z_star), -0.5));
+	grad_potential[2] += (z / std::abs(z)) * FastMath::pow10(ginterp);
```

Comment:
As I mentioned above, if you used `FastMath::log10` to compute the tabulated values, then this is fine, since that is its intended application. If you used `np.log10`, then this gives the wrong answer.

Reply:
I think it's best to use `std::pow(10, ginterp)` for now.

In the future, you could re-calculate the tabulated values with `FastMath::log10` and then go back to using `FastMath::pow10` here.